src/AstArgument.h

```diff
@@ -356,18 +367,43 @@ class AstUserDefinedFunctor : public AstFunctor {
         return name;
     }
 
-    /** set name */
-    void setName(const std::string& name) {
-        this->name = name;
+    /** get type of the functor argument*/
+    RamTypeAttribute getArgType(const size_t arg) const override {
+        return argTypes.at(arg);
+    }
+
+    /** get type of the functor argument*/
+    RamTypeAttribute getReturnType() const override {
+        return returnType;
+    }
+
+    void setArgsTypes(std::vector<RamTypeAttribute> types) {
+        assert(types.size() == args.size() && "Size of types must match size of arguments");
+        argTypes = types;
+    }
+
+    const std::vector<RamTypeAttribute>& getArgsTypes(void) const {
+        return argTypes;
+    }
+
+    void setReturnType(RamTypeAttribute type) {
+        returnType = type;
     }
 
     AstUserDefinedFunctor* clone() const override {
-        auto res = new AstUserDefinedFunctor();
-        for (auto& cur : args) {
-            res->args.emplace_back(cur->clone());
+        auto res = new AstUserDefinedFunctor(name);
+        // Set args
+        for (auto& arg : args) {
+            res->args.emplace_back(arg->clone());
+        }
+        // Set types
+        // Only copy types if they have already been set.
+        if (argTypes.size() > 0) {
```

Comment:
Use `!empty()` rather than `size() > 0`:
```suggestion
        if (!argTypes.empty()) {
```